### PR TITLE
V1.3 Pipeline refactor

### PR DESCRIPTION
**Pull request by chuan-wang**

### CODE REVIEW

**Comment by ewels** [comment]
Great! Some of your file renaming will break things though, so please make these changes (mostly back to what they were?)
- `.gitattributes.txt` back to `.gitattributes`
- `.gitignore.txt` back to `.gitignore`
- `LICENSE.md.txt` to `LICENCE.md`

I'm afraid to say that I will probably replace all of this again soon, as much of this refactored code has been refactored a second time. But if this code is tested and working as it is, I'm happy to merge this now.

**Comment by chuan-wang** [comment]
They have been renamed. Thanks! /chuan

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,3 @@
 .nextflow*
 work/
 data/
-.DS_Store
```
Comment (ewels) [inline .gitignore:4]: Why did you remove this line? This is a common hidden file added on OSX systems, we don't want to accidentally add it to the repo.

File: conf/uppmax.config [hunk 2]
```diff
@@ -13,8 +16,36 @@ process {
...
-    '-A <PROJECT>'
+    "-A $params.project"
   }
+
+  errorStrategy = task.exitStatus == 143 ? 'retry' : 'ignore'
+  maxRetries = 3
+  maxErrors = '-1'
+
+  // Module specific retrying
+  $fastqc.errorStrategy = task.exitStatus == 143 ? 'retry' : 'warning'
```
Comment (ewels) [inline conf/uppmax.config:27]: There are a bunch of errors here that were mine in the RNA pipeline, that I now know about. I'll replace this when I do the new refactoring again shortly though.

For reference, this should be `$fastqc.errorStrategy = { task.exitStatus == 143 ? 'retry' : 'warning' }`. It has to be inside `{}` if you want it to be evaluated at run time. You need this, as `task.exitStatus` doesn't exist until the process has run.

File: main.nf [hunk 3]
```diff
@@ -98,11 +97,9 @@ log.info "R libraries          : ${params.rlocation}"
 log.info "Script dir           : $baseDir"
 log.info "Working dir          : $workDir"
 log.info "Output dir           : ${params.outdir}"
+log.info "Config Profile : ${workflow.profile}"
```
Comment (ewels) [inline main.nf:100]: Minor thing, but it's nice to have the `:` lined up with the others above it.

File: main.nf [hunk 4]
```diff
@@ -117,24 +114,23 @@ else{
...
-// if( !hairpin.exists() ) exit 1, "Missing Bowtie 1 miRBase_hairpin index: ${hairpin}"
-// if( !index.exists() ) exit 1, "Missing Bowtie 2 index: ${index}"
-// if( !gtf.exists() )   exit 2, "Missing GTF annotation: ${gtf}"
-// if( !bed12.exists() ) exit 2, "Missing BED12 annotation: ${bed12}"
+// Validate inputs
+//if( !mature.exists() ) exit 1, "Missing Bowtie 1 miRBase_mature index: ${mature}"
+//if( !hairpin.exists() ) exit 1, "Missing Bowtie 1 miRBase_hairpin index: ${hairpin}"
+//if( !index.exists() ) exit 1, "Missing Bowtie 2 index: ${index}"
+//if( !gtf.exists() )   exit 2, "Missing GTF annotation: ${gtf}"
+//if( !bed12.exists() ) exit 2, "Missing BED12 annotation: ${bed12}"
```
Comment (ewels) [inline main.nf:122]: I will rewrite this when I go through this pipeline later. Will definitely be possible to get it to work.

**Comment by ewels** [comment]
I'll merge now and then go through and make changes myself if that's ok. No need for us to duplicate the same work.

File: .gitignore (hunk 1 above)
Comment (chuan-wang) [reply .gitignore:4]: I copied this file from the NGI-ChIPseq pipeline. Maybe we need to add it there also?

File: conf/uppmax.config (hunk 2 above)
Comment (chuan-wang) [reply conf/uppmax.config:27]: The same as the NGI-ChIP-seq pipeline.

File: main.nf (hunk 4 above)
Comment (chuan-wang) [reply main.nf:122]: Please note that for mature miRNA and hairpin, index, gtf and bed12 are not used.

Comment (chuan-wang) [reply main.nf:122]: And if the alignment is to be done against miRNA of all species (ALL option), index, gtf and bed12 are not used at all in the pipeline.

**Comment by chuan-wang** [comment]
Thanks phil! Please just let me know if there is anything I can do.

**Comment by ewels** [comment]
Yup! Will put in a PR soon, would be good if you could test it :) Or point me towards some data where I can test it myself..

**Comment by chuan-wang** [comment]
I have some test data and it can be done really fast. Just let me know when
the PR is ready.
